mycpp/gc_str.cc

```diff
@@ -462,3 +468,108 @@ List<Str*>* Str::split(Str* sep) {
 
   return result;
 }
+
+static inline Str* _StrFormat(const char* fmt, int fmt_len, va_list args) {
+  auto beg = std::cregex_iterator(fmt, fmt + fmt_len, gStrFmtRegex);
+  auto end = std::cregex_iterator();
+
+  char int_buf[kMaxFmtWidth];
+  std::string buf;
+  for (std::cregex_iterator it = beg; it != end; ++it) {
+    const std::cmatch& match = *it;
+
+    const std::csub_match& lit_m = match[1];
+    assert(lit_m.matched);
+    const std::string& lit_s = lit_m.str();
+    buf.append(lit_s);
+
+    int width = 0;
+    bool zero_pad = false;
+    const std::csub_match& width_m = match[2];
+    const std::string& width_s = width_m.str();
+    if (width_m.matched && !width_s.empty()) {
+      if (width_s[0] == '0') {
+        zero_pad = true;
+        assert(width_s.size() > 1);
+        assert(StringToInteger(width_s.c_str() + 1, width_s.size() - 1, 10,
+                               &width));
+      } else {
+        assert(StringToInteger(width_s.c_str(), width_s.size(), 10, &width));
+      }
+      assert(width >= 0 && width < kMaxFmtWidth);
+    }
+
+    char const* str_to_add = nullptr;
+    int add_len = 0;
+    const std::csub_match& code_m = match[3];
+    const std::string& code_s = code_m.str();
+    if (!code_m.matched) {
```

Comment:
This part confused me for awhile ...  

I guess one problem with the asserts is that it's hard to test the error cases, hm.  

I just checked C's printf is that it doesn't seem to have error cases!  It just gives you garbage :-P

Hm actually can we just get rid of the goto and make this `break` ?

Because nothing in the `done:` block will get executed?

Reply:
Oh you're right! This `goto` is completely unnecessary, since the literal part is appended directly to `buf`. Pushed a commit to remove it